core/ir/isa_regdeps/encoding_common.h

```diff
@@ -0,0 +1,126 @@
+ *   encoded instruction because they are always read;
+ * - 4 bits, #dst: number of destination operands (written) that are registers.
+ *
+ * We assume these encoded values to be little-endian.  Note that we are only interested
+ * in register dependencies, hence operands that are not registers, such as immediates or
+ * memory references, are not present.
+ *
+ * Following the 4 header bytes are the bytes for the operation size and for the encoding
+ * of register operands, if any are present.
+ *
+ * The first byte contains the operation size encoded as an OPSZ_ enum value.  The
+ * operation size is the size of the largest source operand, regardless of it being a
+ * register, a memory reference, or an immediate.
+ *
+ * Following the operation size are the register operand IDs.  Each register operand is 1
+ * byte.  The destination operands go first, followed by the source operands.  An
+ * instruction can have up to 8 operands (sources + destinations).  Note that, because of
+ * 4 byte alignment, the length of encoded instructions will include padding and is as
+ * follows:
+ * - instructions with no operands have only the 4 header bytes (no operation size byte
+ *   nor operand-related bytes);
+ * - instructions with 1 to 3 operands have a length of 8 bytes (4 header bytes + 1 byte
+ *   for operation size + 3 operand-related/padding bytes);
+ * - instructions with 4 to 7 operands have a length of 12 bytes;
+ * - instructions with 8 operands have the maximum length of 16 bytes.
+ *
+ * For example, an instruction with 4 operands (1 dst, 3 src) has a length of 12 bytes and
+ * would be encoded as:
+ * |----------------------| |--| |----| |----|
+ * 31..               ..10  9,8   7..4   3..0
+ *        category         eflags #src   #dst
+ * |--------| |--------| |--------| |--------|
+ * 31.. ..24  23.. ..16  15..  ..8  7..   ..0
+ *  src_op1    src_op0    dst_op0    op_size
+ * |--------| |--------| |--------| |--------|
+ * 31.. ..24  23.. ..16  15..  ..8  7..   ..0
+ *  padding    padding    padding    src_op2
+ *
+ * Because of 4 byte alignment, the last 3 bytes [31.. ..8] are padding and are undefined
+ * (i.e., it cannot be assumed that they have been zeroed-out or contain any meaningful
+ * value).
+ */
+
+#define CATEGORY_BITS 22
+#define FLAGS_BITS 2
+#define NUM_OPND_BITS 4
+
+#define SRC_OPND_SHIFT NUM_OPND_BITS
+#define FLAGS_SHIFT (2 * NUM_OPND_BITS)
+#define CATEGORY_SHIFT (2 * NUM_OPND_BITS + FLAGS_BITS)
+
+#define DST_OPND_MASK ((1U << NUM_OPND_BITS) - 1)
+#define SRC_OPND_MASK (((1U << NUM_OPND_BITS) - 1) << SRC_OPND_SHIFT)
+#define FLAGS_MASK (((1U << FLAGS_BITS) - 1) << FLAGS_SHIFT)
+#define CATEGORY_MASK (((1U << CATEGORY_BITS) - 1) << CATEGORY_SHIFT)
+
+#define REGDEPS_INSTR_WRITES_ARITH 0x1
+#define REGDEPS_INSTR_READS_ARITH 0x2
+
+#define HEADER_BYTES 4
+#define OP_SIZE_INDEX HEADER_BYTES
+#define OPND_INDEX OP_SIZE_INDEX + 1
+
+#define ALIGN_BYTES 4
```

Comment:
Might be safest to put a prefix on all these to avoid collisions: REGDEPS_ or sthg.

Reply:
Done.